internal/cmd/agent/controller/bundledeployment_controller.go

```diff
@@ -0,0 +1,241 @@
+	status, err := r.Deployer.DeployBundle(ctx, bd)
+	if err != nil {
+		logger.V(1).Error(err, "Failed to deploy bundle", "status", status)
+
+		// do not use the returned status, instead set the condition and possibly a timestamp
+		bd.Status = setCondition(bd.Status, err, condition.Cond(fleetv1.BundleDeploymentConditionDeployed))
+
+		merr = append(merr, fmt.Errorf("failed deploying bundle: %w", err))
+	} else {
+		bd.Status = setCondition(status, nil, condition.Cond(fleetv1.BundleDeploymentConditionDeployed))
+	}
+
+	// if we can't retrieve the resources, we don't need to try any of the other operations and requeue now
+	resources, err := r.Deployer.Resources(bd.Name, bd.Status.Release)
+	if err != nil {
+		logger.V(1).Info("Failed to retrieve bundledeployment's resources")
+		// merr = append(merr, fmt.Errorf("failed retrieving resources: %w", err))
+		if err := updateStatus(ctx, logger, r.Client, req.NamespacedName, bd.Status); err != nil {
+			merr = append(merr, fmt.Errorf("failed to update the status: %w", err))
+		}
+		return ctrl.Result{}, errutil.NewAggregate(merr)
+	}
+
+	var result ctrl.Result
+	if monitor.ShouldUpdateStatus(bd) {
+		// update the bundledeployment status and check if we deploy an agent, or if we need to trigger drift correction
+		status, err = r.Monitor.UpdateStatus(ctx, bd, resources)
+		if err != nil {
+			logger.Error(err, "Cannot monitor deployed bundle")
+
+			// if there is an error, do not use the returned status from monitor
+			bd.Status = setCondition(bd.Status, err, condition.Cond(fleetv1.BundleDeploymentConditionMonitored))
+			merr = append(merr, fmt.Errorf("failed updating status: %w", err))
+		} else {
+			// we add to the status from deployer.DeployBundle
+			bd.Status = setCondition(status, nil, condition.Cond(fleetv1.BundleDeploymentConditionMonitored))
+		}
+
+		// Run drift correction
+		if len(status.ModifiedStatus) > 0 && bd.Spec.CorrectDrift.Enabled {
+			err = r.Deployer.RemoveExternalChanges(ctx, bd)
+			if err != nil {
+				merr = append(merr, fmt.Errorf("failed reconciling drift: %w", err))
+			}
+		}
+
+		// TODO is this needed with drift correction?
+		if len(bd.Status.ModifiedStatus) > 0 && monitor.ShouldRedeploy(bd) {
+			result.RequeueAfter = durations.MonitorBundleDelay
+			logger.Info("Redeploying, by resetting the appliedDeploymentID")
+			bd.Status.AppliedDeploymentID = ""
+
+			if monitor.IsAgent(bd) {
+				if err := r.Cleanup.OldAgent(ctx, status.ModifiedStatus); err != nil {
+					merr = append(merr, fmt.Errorf("failed cleaning old agent: %w", err))
+				}
+			}
+		}
+	}
+
+	// update our mini controller, which watches deployed resources for drift
+	_, err = r.DriftDetect.Refresh(logger, req.String(), bd, resources)
+	if err != nil {
+		logger.V(1).Error(err, "Failed to refresh drift detection", "step", "drift")
+		merr = append(merr, fmt.Errorf("failed refreshing drift detection: %w", err))
+	}
+
+	err = r.Cleanup.CleanupReleases(ctx, key, bd)
```

Comment:
We iterate through all BundleDeployments each time any of them changes to see if they need to be cleaned up. That's not very efficient. We can use finalizers as I mentioned in the previous comment

Reply:
Yes, that's old logic again (https://github.com/rancher/fleet/blob/master/internal/cmd/agent/deployer/cleanup/cleanup.go). I'd not fix it within this PR, but I'll add an item to the follow up items.

Reply:
maybe we can fix this and [this](https://github.com/rancher/fleet/pull/1772#discussion_r1407382343) using finalizers in a separate PR?